torchvision/prototype/models/_utils.py

```diff
@@ -1,13 +1,79 @@
+import functools
 import warnings
-from typing import Any, Dict, Optional, TypeVar
+from typing import Any, Dict, Optional, TypeVar, Callable, Tuple, Union
 
-from ._api import WeightsEnum
+from torch import nn
+from torchvision.prototype.utils._internal import kwonly_to_pos_or_kw
 
+from ._api import WeightsEnum
 
 W = TypeVar("W", bound=WeightsEnum)
+M = TypeVar("M", bound=nn.Module)
 V = TypeVar("V")
 
 
+def handle_legacy_interface(**weights: Tuple[str, Union[Optional[W], Callable[[Any, Dict[str, Any]], Optional[W]]]]):
+    """Decorates a model builder with the new interface to make it compatible with the old.
+
+    In particular this handles two things:
+
+    1. Allows positional parameters again, but emits a deprecation warning in case they are used. See
+        :func:`torchvision.prototype.utils._internal.kwonly_to_pos_or_kw` for details.
+    2. Handles the default value change from ``pretrained=False`` to ``weights=None`` and ``pretrained=True`` to
+        ``weights=Weights`` and emits a deprecation warning with instructions for the new interface.
+
+    Args:
+        **weights (Tuple[str, Union[Optional[W], Callable[[Any, Dict[str, Any]], Optional[W]]]]): Deprecated parameter
+            name and default value for the legacy ``pretrained=True``. The default value can be a callable in which
+            case it will be called with the value of ``pretrained`` and the other keyword arguments. For example:
+    """
+
+    def outer_wrapper(builder: Callable[..., M]) -> Callable[..., M]:
+        @kwonly_to_pos_or_kw
+        @functools.wraps(builder)
+        def inner_wrapper(*args: Any, **kwargs: Any) -> M:
+            for weights_param, (pretrained_param, default) in weights.items():  # type: ignore[union-attr]
+                # If neither the weights nor the pretrained parameter as passed, or the weights argument already use
+                # the new style (not a boolean like pretrained did), there is nothing to do.
+                if (weights_param not in kwargs and pretrained_param not in kwargs) or not isinstance(
+                    kwargs.get(weights_param, False), bool
+                ):
+                    continue
+
+                # If the pretrained parameter was passed as positional argument, it is now mapped to
+                # `kwargs[weights_param]`. This happens because the @kwonly_to_pos_or_kw use the current signature to
+                # infer the names of positionally passed arguments and thus has no knowledge that there used to be a
+                # pretrained parameter.
+                pretrained_positional = weights_param in kwargs
+                pretrained_arg = kwargs.pop(weights_param if pretrained_positional else pretrained_param)
+                if pretrained_arg:
+                    default_weights_arg = default(pretrained_arg, kwargs) if callable(default) else default
```

Comment:
I noticed that you changed the lambda to include an extra parameter `pretrained_arg` which you don't currently use. Could you elaborate on which corner case requires that?

Reply:
The `kwargs` argument passed to the callable did not contain the `pretrain_arg`, because it was already removed at that point. Thus, we pass it explicitly to be used in cases where decision for the default weights need to be made with respect to it. See https://github.com/pytorch/vision/pull/5001#discussion_r760989162 for an example.

Reply:
Hmmmm... This makes for a confusing signature. Should we refactor to defer the poping of the param until later to avoid requiring both arguments?

Reply:
I thought about that too, but then we need to decide under what key we put it. Furthermore, if the parameter was passed as positional we need to put it in first to only remove it again after the call.

Finally, what is confusing about

```python
def default(
    pretrained_arg: Any,  # usually bool, but might also be "legacy"
    other_kwargs: Dict[str, Any],
) -> Optional[Weights]:
    ...
```

Reply:
I don't mind adding it and popping it. In a sense you try to decouple all complex scenarios that can occur by collapsing them to 1-2 possible cases. If its positional, you just plug it on the right keyword argument and pop it later no matter its origin (positional or keyword). The only reason I'm discussing this so extensively is because the signature is quite visible from the model builder methods. I wouldn't mind it if that was a technical detail buried somewhere in the decorators. Let me know what you think.

Reply:
Fair point. The internals of the decorator are a little (I promise :stuck_out_tongue: ) more complicated now, but since they are not "visible" that should be fine.